File: src/cz/vutbr/fit/interlockSim/objects/cells/OrientedNodeCell.java
/* Brno University of Technology
 * Faculty of Information Technology
 * 
 * BSc Thesis	2006/2007
 * 
 * Railway Interlocking Simulator
 * 
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.objects.cells;

import java.util.EnumSet;
import java.util.Set;

import cz.vutbr.fit.interlockSim.objects.paths.OrientedPathSeparator;

/**
 * Base implementation of {@link OrientedPathSeparator}
 */
public abstract class OrientedNodeCell extends NodeCell implements OrientedPathSeparator {
	private final boolean orientation;

	protected OrientedNodeCell(boolean orientation, SpatialType spatialType) {
		super(spatialType);
		this.orientation = orientation;
	}

	public Set<Segment> possibleFollowers(Segment from) {
		return EnumSet.of(getFollowingSegment(from));
	}
	
	public boolean getOrientation() {
		return orientation;
	}
	
	public Segment direction() {
		SpatialType st = getSpatialType(); assert st != null : this;
		return st.getSegments()[getOrientation()? 1 : 0];
	}
}

File: src/main/kotlin/cz/vutbr/fit/interlockSim/objects/cells/OrientedNodeCell.kt
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.objects.cells

import cz.vutbr.fit.interlockSim.exceptions.requireSimulationNotNull
import cz.vutbr.fit.interlockSim.objects.core.Cell
import cz.vutbr.fit.interlockSim.objects.core.OrientedPathSeparator
import java.util.EnumSet

/**
 * Base implementation of {@link OrientedPathSeparator}
 */
abstract class OrientedNodeCell protected constructor(
	private val orientation: Boolean,
	spatialType: Cell.SpatialType
) : NodeCell(spatialType),
	OrientedPathSeparator {
	/**
	 * Get the following segment from a given segment (static configuration).
	 * This is used by possibleFollowers and also needed by Dynamic wrappers.
	 */
	abstract fun getFollowingSegment(from: Cell.Segment?): Cell.Segment?

	override fun possibleFollowers(from: Cell.Segment): Set<Cell.Segment> =
		EnumSet.of(getFollowingSegment(from)) as Set<Cell.Segment>

	override fun getOrientation(): Boolean = orientation

	override fun direction(): Cell.Segment {
		val st = getSpatialType()
		requireSimulationNotNull(st) { "Spatial type must not be null for: $this" }
		return st.segments[if (getOrientation()) 1 else 0]
	}
}
